File: src/main/resources/static/ts/PhysicsState.ts
import Vector = require('./Vector');
import Rect = require('./Rect');
import Collision = require('./Collision');
import GameObject = require('./GameObjects/GameObject');

export = class PhysicsState {
    static gravityForce: number;

    gameObject: GameObject;
    gravityForce: number;
    gravity: boolean;
    still: boolean;
    colliders: GameObject[];
    velocity: Vector;

    constructor(gameObject: GameObject) {
        this.gameObject = gameObject;
        this.gravityForce = PhysicsState.gravityForce;
        this.gravity = false;
        this.still = true;
        this.colliders = [];
        this.velocity = Vector.zero();
    };

    change(dt: number) {
        if (this.gravity)
            this.velocity.y += this.gravityForce * dt;
        this.gameObject.frame.x += this.velocity.x * dt;
        this.gameObject.frame.y += this.velocity.y * dt;
        if (this.gravity) {
            // game.debugRuler(this.gameObject.frame.x, this.gameObject.frame.y, this.velocity.y/this.gravityForce*100);
            // console.log(this.velocity.y);
            // console.log(this.gameObject.frame.y);
            // console.log(dt);
        }
    };

    detectCollision(c: PhysicsState, dt: number) {
        if (this.still && c.still)
            return;//two still objects cannot collide (because we presume that they both do not move)

        let alreadyCollided = false;

        let i, len;
        i = 0;
        len = this.colliders.length;
        while (i < len) {
            if (this.colliders[i] === c.gameObject)
                alreadyCollided = true;
            i++;
        }
        if (!alreadyCollided) {
            i = 0;
            len = c.colliders.length;
            while (i < len) {
                if (c.colliders[i] === this.gameObject)
                    alreadyCollided = true;
            }
        }

        const center1 = this.gameObject.getGlobalPosition();
        const center2 = c.gameObject.getGlobalPosition();
        const overlapArea = Rect.overlapArea(
            center1.x, center1.y, this.gameObject.frame.width, this.gameObject.frame.height,
            center2.x, center2.y, c.gameObject.frame.width, c.gameObject.frame.height
        );
        if (overlapArea) {
            if (!alreadyCollided) {
                this.addCollider(c);
                this.enterCollision(c, overlapArea, dt);
            }
            this.collision(c, overlapArea, dt);
        } else if (alreadyCollided) {
            this.removeCollider(c);
            this.exitCollision(c, dt);
        }

        if (this.gameObject.isRemoved() || c.gameObject.isRemoved())
            this.removeCollider(c);
    };

    addCollider(collider: PhysicsState) {
        this.colliders.push(collider.gameObject);
        collider.colliders.push(this.gameObject);
    };

    removeCollider(collider: PhysicsState) {
        this.colliders.splice(this.colliders.indexOf(collider.gameObject), 1);
        collider.colliders.splice(this.colliders.indexOf(this.gameObject), 1);
    };

    collision(collider: PhysicsState, overlapArea: Vector, dt: number) {
        this.gameObject.handleCollision(new Collision(collider.gameObject, overlapArea), dt);
        collider.gameObject.handleCollision(new Collision(this.gameObject, overlapArea.copy().multiply(-1)), dt);
    };

    enterCollision(collider: PhysicsState, overlapArea: Vector, dt: number) {
        this.gameObject.handleEnterCollision(new Collision(collider.gameObject, overlapArea), dt);
        collider.gameObject.handleEnterCollision(new Collision(this.gameObject, overlapArea.copy().multiply(-1)), dt);
    };

    exitCollision(collider: PhysicsState, dt: number) {
        this.gameObject.handleExitCollision(collider.gameObject, dt);
        collider.gameObject.handleExitCollision(this.gameObject, dt);
    };
}

File: src/main/kotlin/me/demoniacdeath/onlineplatformer/PhysicsState.kt
package me.demoniacdeath.onlineplatformer

import com.fasterxml.jackson.annotation.JsonIgnore

class PhysicsState (@JsonIgnore val gameObject: GameObject) {
    var velocity: Vector = Vector()
    var gravity: Boolean = false
    var still: Boolean = true
    var gravityForce: Double = 0.0
    val colliders: MutableSet<GameObject> = HashSet()

    fun change() {
        if (gravity) {
            velocity += Vector(0.0, gravityForce)
        }
        gameObject.frame.center += velocity
    }

    fun clean() {
        colliders.removeAll { it.removed }
    }

    fun detectCollisions(c: PhysicsState?) {
        if (c === null) return
        if (still && c.still) return

        val x1 = gameObject.globalPosition().x - gameObject.frame.size.width / 2
        val x2 = c.gameObject.globalPosition().x - c.gameObject.frame.size.width / 2
        val X1 = x1 + gameObject.frame.size.width
        val X2 = x2 + c.gameObject.frame.size.width
        val y1 = gameObject.globalPosition().y - gameObject.frame.size.height / 2
        val y2 = c.gameObject.globalPosition().y - c.gameObject.frame.size.height / 2
        val Y1 = y1 + gameObject.frame.size.height
        val Y2 = y2 + c.gameObject.frame.size.height

        val diffX1 = X1 - x2
        val diffX2 = x1 - X2
        val diffY1 = Y1 - y2
        val diffY2 = y1 - Y2

        val alreadyCollided = colliders.contains(c.gameObject) || c.colliders.contains(gameObject)
        if (diffX1 > 0 &&
                diffX2 < 0 &&
                diffY1 > 0 &&
                diffY2 < 0) {
            val overlapArea = Vector(
                    if (Math.abs(diffX1) < Math.abs(diffX2)) diffX1 else diffX2,
                    if (Math.abs(diffY1) < Math.abs(diffY2)) diffY1 else diffY2
            )
            if (!alreadyCollided) {
                colliders.add(c.gameObject)
                c.colliders.add(gameObject)

                gameObject.handleEnterCollision(Collision(c.gameObject, overlapArea))
                c.gameObject.handleEnterCollision(Collision(gameObject, overlapArea * -1.0))
            }
            gameObject.handleCollision(Collision(c.gameObject, overlapArea))
            c.gameObject.handleCollision(Collision(gameObject, overlapArea * -1.0))
        } else if (alreadyCollided) {
            colliders.remove(c.gameObject)
            c.colliders.remove(gameObject)
            gameObject.handleExitCollision(c.gameObject)
            c.gameObject.handleExitCollision(gameObject)
        }
    }

}
